Add ConfluentKafka instrumentation

g7ed6e opened the pull request:
Fixes #1485.

## Changes

Add an instrumentation package for Confluent.Kafka library.
Implement tracing and propagation through Kafka message headers
Implement metrics 

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

eerhardt commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/AssemblyInfo.cs:
```diff
@@ -0,0 +1,21 @@
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("OpenTelemetry.Instrumentation.ConfluentKafka.Tests" + AssemblyInfo.PublicKey)]
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2" + AssemblyInfo.MoqPublicKey)]
```
Is this being used? I don't see Moq being used anywhere in the PR (but I might be missing something).

g7ed6e replied:
dropped. Moq has been removed since I opened the PR, i adapted the unit tests but forgot this. thanks

eerhardt commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetry.Instrumentation.ConfluentKafka.csproj:
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>net8.0;netstandard2.0</TargetFrameworks>
```
Why not `net6.0`?

g7ed6e replied:
added net6.0

eerhardt commented on test/OpenTelemetry.Instrumentation.ConfluentKafka.Tests/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+# Create a container for running the OpenTelemetry Redis integration tests.
```
```suggestion
# Create a container for running the OpenTelemetry ConfluentKafka integration tests.
```

Copy paste leftover.

g7ed6e replied:
done thanks

eerhardt commented on test/OpenTelemetry.Instrumentation.ConfluentKafka.Tests/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+WORKDIR /repo
+COPY . ./
+WORKDIR "/repo/test/OpenTelemetry.Instrumentation.ConfluentKafka.Tests"
+RUN dotnet publish "OpenTelemetry.Instrumentation.ConfluentKafka.Tests.csproj" -c "${PUBLISH_CONFIGURATION}" -f "${PUBLISH_FRAMEWORK}" -o /drop -p:IntegrationBuild=true -p:TARGET_FRAMEWORK=${PUBLISH_FRAMEWORK}
```
What does `/p:IntegrationBuild=true` do?

g7ed6e replied:
It avoids taking a private dependency on `MinVer`. See https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/e42f6cdda1120e7630aa69e0ab7e8560feddba31/build/Common.prod.props#L23

eerhardt commented on test/OpenTelemetry.Instrumentation.ConfluentKafka.Tests/docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+# Start a kafka container and then run OpenTelemetry redis integration tests.
```
```suggestion
# Start a kafka container and then run OpenTelemetry kafka integration tests.
```

eerhardt commented:
On the Tracing side, would it be better if we just finished up https://github.com/confluentinc/confluent-kafka-dotnet/pull/1838 (and also made a similar PR for Consumer)? That way we wouldn't need to wrap IProducer / IConsumer. The OpenTelemetry package would just enable the ActivitySource from the Kafka library.

cc @cijothomas

g7ed6e commented:
> > On the Tracing side, would it be better if we just finished up [confluentinc/confluent-kafka-dotnet#1838](https://github.com/confluentinc/confluent-kafka-dotnet/pull/1838) (and also made a similar PR for Consumer)? That way we wouldn't need to wrap IProducer / IConsumer. The OpenTelemetry package would just enable the ActivitySource from the Kafka library.
> 
> > 
> 
> > cc @cijothomas
> 
> 
> 
> Yes. If that PR finishes, then there may not be even a need of an instrumentation library here, as users can just add AddSource themselves.

@eerhardt @cijothomas I agree but I think until that it would be great to get tracing support.

cijothomas commented:
> > > On the Tracing side, would it be better if we just finished up [confluentinc/confluent-kafka-dotnet#1838](https://github.com/confluentinc/confluent-kafka-dotnet/pull/1838) (and also made a similar PR for Consumer)? That way we wouldn't need to wrap IProducer / IConsumer. The OpenTelemetry package would just enable the ActivitySource from the Kafka library.
> > 
> > 
> > > 
> > 
> > 
> > > cc @cijothomas
> > 
> > 
> > Yes. If that PR finishes, then there may not be even a need of an instrumentation library here, as users can just add AddSource themselves.
> 
> @eerhardt @cijothomas I agree but I think until that it would be great to get tracing support.

I am not quite sure if the efforts of maintaining a library is worth it, unless we know that the native tracing support will take a very long time. If it is coming in few months, wouldn't that be better to wait?

g7ed6e commented:
> > > > On the Tracing side, would it be better if we just finished up [confluentinc/confluent-kafka-dotnet#1838](https://github.com/confluentinc/confluent-kafka-dotnet/pull/1838) (and also made a similar PR for Consumer)? That way we wouldn't need to wrap IProducer / IConsumer. The OpenTelemetry package would just enable the ActivitySource from the Kafka library.
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > cc @cijothomas
> 
> > > 
> 
> > > 
> 
> > > Yes. If that PR finishes, then there may not be even a need of an instrumentation library here, as users can just add AddSource themselves.
> 
> > 
> 
> > @eerhardt @cijothomas I agree but I think until that it would be great to get tracing support.
> 
> 
> 
> I am not quite sure if the efforts of maintaining a library is worth it, unless we know that the native tracing support will take a very long time. If it is coming in few months, wouldn't that be better to wait?

@cijothomas Having a look back to the original issue https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269 I notice PRs to auto instrument Kafka has been linked here https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2854. I think it could make sense to offer a per producer / per consumer optin approach that fit dotnet/aspire integration https://github.com/dotnet/aspire/pull/951 which is the primary motivation of this contribution. Notice that implementation from auto instrumentation and the current PR are not aligned but it could be easily done to offer consistency.

g7ed6e commented:
> > > > > On the Tracing side, would it be better if we just finished up [confluentinc/confluent-kafka-dotnet#1838](https://github.com/confluentinc/confluent-kafka-dotnet/pull/1838) (and also made a similar PR for Consumer)? That way we wouldn't need to wrap IProducer / IConsumer. The OpenTelemetry package would just enable the ActivitySource from the Kafka library.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > > cc @cijothomas
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > Yes. If that PR finishes, then there may not be even a need of an instrumentation library here, as users can just add AddSource themselves.
> > 
> > 
> > > 
> > 
> > 
> > > @eerhardt @cijothomas I agree but I think until that it would be great to get tracing support.
> > 
> > 
> > I am not quite sure if the efforts of maintaining a library is worth it, unless we know that the native tracing support will take a very long time. If it is coming in few months, wouldn't that be better to wait?
> 
> @cijothomas Having a look back to the original issue [confluentinc/confluent-kafka-dotnet#1269](https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269) I notice PRs to auto instrument Kafka has been linked here [open-telemetry/opentelemetry-dotnet-instrumentation#2854](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2854). I think it could make sense to offer a per producer / per consumer optin approach that fit dotnet/aspire integration [dotnet/aspire#951](https://github.com/dotnet/aspire/pull/951) which is the primary motivation of this contribution. Notice that implementation from auto instrumentation and the current PR are not aligned but it could be easily done to offer consistency.

Hi @cijothomas. May I ask for a review ? What do you think about maintaining this code until official support comes out, then the library can be deprecated ?

pyohannes commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,183 @@
+    {
+        if (this.options.Traces)
+        {
+            using Activity? activity = this.StartActivity("publish", topicPartition) ?? Activity.Current;
+            this.InjectActivity(activity, message);
+        }
+
+        return await this.producerImplementation.ProduceAsync(topicPartition, message, cancellationToken).ConfigureAwait(false);
```
This activity should measure the duration of the `ProduceAsync` call. it should be disposed or stopped after `ProduceAsync` was awaited.

g7ed6e replied:
done

pyohannes commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,183 @@
+    {
+        if (this.options.Traces)
+        {
+            using Activity? activity = this.StartActivity("publish", topic) ?? Activity.Current;
+            this.InjectActivity(activity, message);
+        }
+
+        this.producerImplementation.Produce(topic, message, deliveryHandler);
```
Same here. The activity should be stopped or disposed after `Produce` returned.

pyohannes commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,183 @@
+    {
+        if (this.options.Traces)
+        {
+            using Activity? activity = this.StartActivity("publish", topic) ?? Activity.Current;
+            this.InjectActivity(activity, message);
+        }
+
+        this.producerImplementation.Produce(topic, message, deliveryHandler);
```
If this throws, you should set the status of the started activity to `ERROR` and also set `error.type`.

g7ed6e replied:
done

pyohannes commented on examples/Examples.ConfluentKafka/Program.cs:
```diff
@@ -0,0 +1,95 @@
+using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+    .AddConsoleExporter()
+    .AddZipkinExporter()
+    .AddKafkaProducerInstrumentation(producerBuilder)
+    .AddKafkaConsumerInstrumentation("consumer-a", consumerBuilderA)
+    .AddKafkaConsumerInstrumentation("consumer-b", consumerBuilderB)
```
This is a bit cumbersome, as it requires 3 sequential initialization steps:

1. Initialize a consumer/producer builder.
2. Initialize the OTel SDK.
3. Initialize the consumer/producer.

Probably not a showstopper, but feels a bit weird to have that three-way DI dependency.

pyohannes commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/ConfluentKafkaMeterInstrumentation.cs:
```diff
@@ -0,0 +1,194 @@
+                    return;
+                }
+
+                TagList tags = new()
+                {
+                    { Tags.ClientId, statistics.ClientId },
+                    { Tags.Name, statistics.Name },
+                };
```
To be in conformance with [messaging semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-metrics.md), there should be a tag for `messaging.system`, and for `messaging.destination.name` and `server.address` if those are available.

pyohannes reviewed (commented):
Thanks for submitting this. I'm reviewing mostly from the point of view of a contributor to messaging semantic conventions.

g7ed6e commented:
> Thanks for submitting this. I'm reviewing mostly from the point of view of a contributor to messaging semantic conventions.

Hi @pyohannes Thank you for your review. I should be able to get back to this and incorporate your suggestions next week.

cijothomas commented:
> > > > > > On the Tracing side, would it be better if we just finished up [confluentinc/confluent-kafka-dotnet#1838](https://github.com/confluentinc/confluent-kafka-dotnet/pull/1838) (and also made a similar PR for Consumer)? That way we wouldn't need to wrap IProducer / IConsumer. The OpenTelemetry package would just enable the ActivitySource from the Kafka library.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > > cc @cijothomas
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Yes. If that PR finishes, then there may not be even a need of an instrumentation library here, as users can just add AddSource themselves.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > @eerhardt @cijothomas I agree but I think until that it would be great to get tracing support.
> > > 
> > > 
> > > I am not quite sure if the efforts of maintaining a library is worth it, unless we know that the native tracing support will take a very long time. If it is coming in few months, wouldn't that be better to wait?
> > 
> > 
> > @cijothomas Having a look back to the original issue [confluentinc/confluent-kafka-dotnet#1269](https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269) I notice PRs to auto instrument Kafka has been linked here [open-telemetry/opentelemetry-dotnet-instrumentation#2854](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2854). I think it could make sense to offer a per producer / per consumer optin approach that fit dotnet/aspire integration [dotnet/aspire#951](https://github.com/dotnet/aspire/pull/951) which is the primary motivation of this contribution. Notice that implementation from auto instrumentation and the current PR are not aligned but it could be easily done to offer consistency.
> 
> Hi @cijothomas. May I ask for a review ? What do you think about maintaining this code until official support comes out, then the library can be deprecated ?

Do you know any rough timelines when the official support comes ? If it is in less than 6 months, the maintainence+deprecation overhead here won't feel like worth.. If its an year or so at least before official support comes, it is okay to keep this PR moving.

vishweshbankwar commented:
@g7ed6e Is it possible for you to follow up with owners on the status of https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269 once more? Having the instrumentation supported natively is the ideal scenario and many libraries are moving towards that. If the feature is not planned then we can consider adding this instrumentation here.

Note that you would need to follow the [process outlined](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md#contributing-a-new-project) to get started. Also could you clarify who will be the owner(s) for this library. 

cc: @open-telemetry/dotnet-contrib-maintainers

g7ed6e commented:
@vishweshbankwar / @open-telemetry/dotnet-contrib-maintainers

May i ask you to reopen this PR ? I'm currently still following up the status with owners.

g7ed6e commented:
@vishweshbankwar @cijothomas latest move in https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269 states that tracing won't be covered. Should we reconsider adding tracing through this repo?

vishweshbankwar commented:
> @vishweshbankwar @cijothomas latest move in [confluentinc/confluent-kafka-dotnet#1269](https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269) states that tracing won't be covered. Should we reconsider adding tracing through this repo?

@g7ed6e - Thanks for following up. 

> Should we reconsider adding tracing through this repo?

Sounds good to me.

g7ed6e commented:
> > @vishweshbankwar @cijothomas latest move in [confluentinc/confluent-kafka-dotnet#1269](https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269) states that tracing won't be covered. Should we reconsider adding tracing through this repo?
> 
> @g7ed6e - Thanks for following up.
> 
> > Should we reconsider adding tracing through this repo?
> 
> Sounds good to me. Assuming you would be owing this package?

Thanks @vishweshbankwar. I rebased the branch and joined the slack otel-dotnet channel.

Kielek commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/.publicApi/net6.0/PublicAPI.Shipped.txt:
Please apply changes from #1711 also to this project.
If public api is common, move one copy of the files directly to .publicApi folder, remove all copies in subfolders.

Kielek commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetry.Instrumentation.ConfluentKafka.csproj:
Please add README.md and CHANGELOG.md (just scaffolding with unreleased section)).

g7ed6e replied:
done

Kielek commented on build/Projects/OpenTelemetry.Instrumentation.ConfluentKafka.proj:
Please extend also `.github/workflows/ci.yml`.

g7ed6e replied:
done

Kielek commented on build/Projects/OpenTelemetry.Instrumentation.ConfluentKafka.proj:
Please add yourself as copmonent owner to  .github/component_owners.yml

g7ed6e replied:
done

Kielek commented on examples/Examples.ConfluentKafka/Examples.ConfluentKafka.csproj:
```diff
@@ -0,0 +1,15 @@
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="OpenTelemetry.Exporter.Zipkin" Version="$(OpenTelemetryCoreLatestVersion)" />
+    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.3.1" />
```
```suggestion
    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="$(OpenTelemetryCoreLatestVersion)" />
```

g7ed6e replied:
done

Kielek commented on examples/Examples.ConfluentKafka/Examples.ConfluentKafka.csproj:
```diff
@@ -0,0 +1,15 @@
+    <ProjectReference Include="..\..\src\OpenTelemetry.Instrumentation.ConfluentKafka\OpenTelemetry.Instrumentation.ConfluentKafka.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="OpenTelemetry.Exporter.Zipkin" Version="$(OpenTelemetryCoreLatestVersion)" />
```
Consider switching to OTLP exporter. It is the standard way to export data.

Kielek commented on build/Projects/OpenTelemetry.Instrumentation.ConfluentKafka.proj:
```diff
@@ -0,0 +1,33 @@
+  <ItemGroup>
+    <SolutionProjects Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.ConfluentKafka\OpenTelemetry.Instrumentation.ConfluentKafka.csproj" />
+    <SolutionProjects Include="$(RepoRoot)\test\OpenTelemetry.Instrumentation.ConfluentKafka.Tests\OpenTelemetry.Instrumentation.ConfluentKafka.Tests.csproj" />
+    <SolutionProjects Include="$(RepoRoot)\examples\redis\**\*.csproj" />
```
I think that you have wrong folder here.
```suggestion
    <SolutionProjects Include="$(RepoRoot)\examples\Examples.ConfluentKafka\Examples.ConfluentKafka.csproj" />
```

g7ed6e replied:
good catch thanks!

Kielek commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetry.Instrumentation.ConfluentKafka.csproj:
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>net8.0;net6.0;netstandard2.0</TargetFrameworks>
```
Why netstandard2.0 is needed? Is it an option to switch in to net462?
```suggestion
    <TargetFrameworks>net8.0;net6.0</TargetFrameworks>
    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);$(NetFrameworkMinimumSupportedVersion)</TargetFrameworks>
```

Kielek commented on test/OpenTelemetry.Instrumentation.ConfluentKafka.Tests/OpenTelemetry.Instrumentation.ConfluentKafka.Tests.csproj:
```diff
@@ -0,0 +1,28 @@
+    <Description>Unit test project for OpenTelemetry ConfluentKafka instrumentation</Description>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>$(SupportedNetTargets)</TargetFrameworks>
+    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);net462</TargetFrameworks>
```
```suggestion
    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);$(NetFrameworkMinimumSupportedVersion)</TargetFrameworks>
```

g7ed6e replied:
done

Kielek commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetry.Instrumentation.ConfluentKafka.csproj:
Please add project to .github/ISSUE_TEMPLATE/bug_report.yml and .github/ISSUE_TEMPLATE/feature_request.yml

g7ed6e replied:
done

Kielek reviewed (commented):
First round of technical review. I do not checked to login in the code.

Kielek commented on .github/workflows/ci.yml:
```diff
@@ -43,6 +43,7 @@ jobs:
           instrumentation-grpcnetclient: ['*/OpenTelemetry.Instrumentation.GrpcNetClient*/**', '!**/*.md']
           instrumentation-hangfire: ['*/OpenTelemetry.Instrumentation.Hangfire*/**', '!**/*.md']
           instrumentation-http: ['*/OpenTelemetry.Instrumentation.Http*/**', '!**/*.md']
+          instrumentation-kafka: ['*/OpenTelemetry.Instrumentation.ConfluentKafka*/**', '!**/*.md']
```
Please use full name of the package, without opentelemetry. prefix.
Also include examples folder

```suggestion
          instrumentation-confluentkafka: ['*/OpenTelemetry.Instrumentation.ConfluentKafka*/**', 'examples/redis/**',  '!**/*.md']
```

Alphabetical reordering + changes in other places in this file is also needed.

g7ed6e replied:
done

Kielek commented on examples/kafka/Examples.ConfluentKafka.csproj:
```diff
@@ -0,0 +1,15 @@
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\OpenTelemetry.Instrumentation.ConfluentKafka\OpenTelemetry.Instrumentation.ConfluentKafka.csproj" />
```
```suggestion
    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.ConfluentKafka\OpenTelemetry.Instrumentation.ConfluentKafka.csproj" />
```

g7ed6e replied:
done

Kielek commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+# Changelog - OpenTelemetry.Instrumentation.ConfluentKafka
+
+## Unreleased
```
```suggestion
# Changelog

## Unreleased

* Initial release
```

g7ed6e replied:
done

Kielek commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/ConfluentKafkaCommon.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    internal static readonly string ActivitySourceName = typeof(ConfluentKafkaInstrumentation).Assembly.GetName().Name!;
+
+    internal static readonly ActivitySource ActivitySource = new(ActivitySourceName);
```
Pass version argument, it can be calculated based on `Assembly.GetPackageVersion()` see AssemblyVersionExtensions class.

g7ed6e replied:
implemented through `ConfluentKafkaCommon.InstrumentationVersion`

g7ed6e commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/ConfluentKafkaMeterInstrumentation.cs:
```diff
@@ -0,0 +1,194 @@
+    public ConfluentKafkaMeterInstrumentation(MetricsChannel channel)
+    {
+        this.channel = channel;
+        this.meter = new(InstrumentationName);
```
implemented through `ConfluentKafkaCommon.InstrumentationVersion`

g7ed6e commented on .github/component_owners.yml:
```diff
@@ -35,6 +35,8 @@ components:
     - ppittle
   src/OpenTelemetry.Instrumentation.Cassandra/:
     - xsoheilalizadeh
+  src/OpenTelemetry.Instrumentation.ConfluentKafka/:
```
done

Kielek commented on .github/workflows/ci.yml:
```diff
@@ -36,6 +36,7 @@ jobs:
           instrumentation-aws: ['*/OpenTelemetry.Instrumentation.AWS/**', '*/OpenTelemetry.Instrumentation.AWS.Tests/**', '!**/*.md']
           instrumentation-aws-lambda: ['*/OpenTelemetry.Instrumentation.AWSLambda/**', '*/OpenTelemetry.Instrumentation.AWSLambda.Tests/**', '!**/*.md']
           instrumentation-cassandra: ['*/OpenTelemetry.Instrumentation.Cassandra*/**', '!**/*.md']
+          instrumentation-confluentkafka: ['*/OpenTelemetry.Instrumentation.ConfluentKafka*/**', '!**/*.md']
```
```suggestion
          instrumentation-confluentkafka: ['*/OpenTelemetry.Instrumentation.ConfluentKafka*/**', 'examples/kafka/**',  '!**/*.md']
```

g7ed6e replied:
done

Kielek commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/README.md:
````diff
@@ -0,0 +1,11 @@
+Download the `OpenTelemetry.Instrumentation.ConfluentKafka` package:
+
+```shell
+dotnet add package OpenTelemetry.Instrumentation.ConfluentKafka
````
This package, for a while, will be not released as stable. Messaging sem. conv. is still not stable.

```suggestion
dotnet add package OpenTelemetry.Instrumentation.ConfluentKafka  --prerelease
```

g7ed6e replied:
done

g7ed6e commented:
@Kielek I'm having issues sharing services (`InstrumentedConsumerBuilder<TKey, TValue>`, `InstrumentedProducerBuilder<TKey, TValue>` and their respective named options) between `IServiceProvider` instances of `MeterProviderBuilder` and `TracerProviderBuilder`.. it looks like each builder has its own `IServiceCollection`, I'm thinking to move `InstrumentedXXXBuilder<TKey, TValue>` instances in static instances of `ConcurrentDictionary<(string name, Type KeyType, Type ValueType), XXXBuilder<TKey, TValue>>`. i will give this a try.

CodeBlanch commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -30,6 +30,7 @@ body:
         - OpenTelemetry.Instrumentation.AWS
         - OpenTelemetry.Instrumentation.AWSLambda
         - OpenTelemetry.Instrumentation.Cassandra
+        - OpenTelemetry.Instrumentation.ConfluentKafka        
```
Please also add in here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/cd7e8d3cb6cb73c9dcb1398ac5914081d7578524/.github/workflows/prepare-release.yml#L22-L23

That list probably wasn't there when you started this work it is a newer thing. Sorry about that!

g7ed6e replied:
done

CodeBlanch commented on .github/workflows/ci.yml:
```diff
@@ -222,6 +223,17 @@ jobs:
+    uses: ./.github/workflows/Component.BuildTest.yml
+    with:
+      project-name: Component[OpenTelemetry.Instrumentation.ConfluentKafka]
+      code-cov-name: Instrumentation.ConfluentKafka          
```
Please define `unittests-Instrumentation.ConfluentKafka` in here:

https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/cd7e8d3cb6cb73c9dcb1398ac5914081d7578524/.github/codecov.yml#L70

Sorry there is a lot of ceremony scattered around 😄

g7ed6e replied:
done

CodeBlanch commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedConsumer.cs:
```diff
@@ -0,0 +1,330 @@
+            : string.Concat(consumeResult.TopicPartitionOffset!.Topic, " ", OperationName);
+
+        using var activity = this.StartReceiveActivity(spanName, propagationContext, startTime, consumeResult.TopicPartitionOffset, consumeResult.Key);
+        activity?.Stop();
```
Will this lead to useful traces?

It will show only a sliver of time for the consumption:

```
Service A (Producer) 

Incoming request _________________________
                             publish___

Service B (Consumer)

                                                  consume_
```

I feel like what users actually want is:

```
Service A (Producer) 

Incoming request _________________________
                             publish___

Service B (Consumer)

                                                  consume_________________________________
                                                             call some other service___
```

Meaning the trace started when a message is consumed should run over everything that happens processing that message, no?

CodeBlanch replied:
@lmolkova Any feel for how traces should be represented when it comes to message queues\pub-sub?

lmolkova replied:
We have messaging semconv with examples here https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md

The short answer - if the user calls `kafkaClient.Poll()` then we should have a receive span. If there is an API in the SDK that allows to do `kafkaClient.Process(msg => {...})`, we should have process span covering the duration of the process.
If user writes

```csharp
var msg = kafkaClient.Poll();
// process msg 
```

then there is nothing instrumentation can do to instrument things after poll call ends. Users are expected to add spans themselves.

CodeBlanch replied:
@g7ed6e Is there any way for the caller consuming the message to access the headers/trace information so they can continue the trace if desired?

g7ed6e replied:
@CodeBlanch Yes. The caller is supposed to call `Consume` in a loop. https://docs.confluent.io/kafka-clients/dotnet/current/overview.html#the-consume-loop.
Each call to `Consume` returns a `ConsumeResult<TKey, TValue>`. Headers are here accessible through the `Message.Headers` property. https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.ConsumeResult-2.html

CodeBlanch replied:
@g7ed6e 

What if the instrumentation provided some extensions to help users do this correctly?

```csharp
namespace Confluent.Kafka;

public static class OpenTelemetryConsumeResultExtensions
{
    public static bool TryExtractPropagationContext<TKey, TValue>(
        this ConsumeResult<TKey, TValue> consumeResult,
        out PropagationContext propagationContext)
    {
        // todo: implement
    }

    public static void ConsumeAndProcessMessage<TKey, TValue>(
        this IConsumer<TKey, TValue> consumer,
        Action<ConsumeResult<TKey, TValue>> processMessage,
        CancellationToken cancellationToken = default)
    {
        var consumeResult = consumer.Consume(cancellationToken);

        if (TryExtractPropagationContext(consumeResult, out var propagationContext))
        {
            // todo: start child span
        }

        try
        {
            processMessage(consumeResult);
        }
        finally
        {
            // todo: stop child span
        }
    }
}
```

CodeBlanch commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,245 @@
+            activity?.Stop();
+        }
+
+        activity?.Stop();
```
Dupe?

g7ed6e replied:
Good catch thanks!

CodeBlanch commented:
@g7ed6e 

> edit: @CodeBlanch is there a way to share services between `MeterProviderBuilder` and `TracerProviderBuilder` ?

The short answer is: No

The long answer is: Depends

For users doing...

```csharp
appBuilder.Services.AddOpenTelemetry()
   .WithTracing(tracing => tracing.AddKafkaConsumerInstrumentation())
   .WithMetrics(metrics => metrics.AddKafkaConsumerInstrumentation());
```

The `IServiceCollection` + `IServiceProvider` will be the one tied to host and everything is shared.

For users doing...

```csharp
using var tracerProvider = Sdk.CreateTracerProviderBuilder()
  .AddKafkaConsumerInstrumentation()
  .Build();

using var meterProvider = Sdk.CreateMeterProviderBuilder()
  .AddKafkaConsumerInstrumentation()
  .Build();
```

They are independent.

Hopefully there will be a new API soon (https://github.com/open-telemetry/opentelemetry-dotnet/pull/5325) for that second case which will also be shared:

```csharp
using var openTelemetry = OpenTelemetrySdk.Create(builder => builder
   .WithTracing(tracing => tracing.AddKafkaConsumerInstrumentation())
   .WithMetrics(metrics => metrics.AddKafkaConsumerInstrumentation()));
```

If you _really_ need them to be singletons then you'll have to do something `static`. If having singletons is just a nice-to-have I would just use the `IServiceProvider` normally and let it sort out what to do.

Remember users can set up multiple pipelines:

```csharp
using var tracerProvider1 = Sdk.CreateTracerProviderBuilder()
  .AddSource("SomeHighPrioritySource")
  .AddKafkaConsumerInstrumentation()
  .AddOtlpExporter(...send one place...)
  .Build();

using var tracerProvider2 = Sdk.CreateTracerProviderBuilder()
  .AddSource("SomeLowPrioritySource")
  .AddKafkaConsumerInstrumentation()
  .AddZipkinExporter(...send some other place...)
  .Build();
```

In which case would having static singletons actually be a problem?

g7ed6e commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedConsumer.cs:
```diff
@@ -0,0 +1,330 @@
+            ? propagationContext.ActivityContext
+            : default;
+
+        Activity? activity = ConfluentKafkaCommon.ActivitySource.StartActivity(spanName, kind: ActivityKind.Consumer, links: activityLinks, startTime: start, parentContext: activityContext);
```
The sample table here https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/kafka.md#apache-kafka-with-quarkus-or-spring-boot-example states the opposite :).
Please advise @lachmatt / @pyohannes / @lmolkova / @Kielek

lmolkova replied:
In case of receive activity, parenting is problematic.
If we could, we should start this activity before receiving starts - we can't do it today because links have to be added at the beginning. But otel spec now allows to add links after span creation and it's a matter of time for .NET to support it.
Starting activity at the beginning allows to correlate any nested spans or logs with it. It won't be possible with the current approach.


For the processing callback instrumentation, having both - links and parenting is just fine. 

I'll file an issue on the example on the semconv - it's not up to date.

g7ed6e replied:
@lmolkova The .NET Kafka SDK exposes a `Subscribe` + `Consume` API that forces a caller to first call `Subscribe` and then call `Consume` in a loop. My understanding of the messaging semantic convention is that each call to `Consume` (which can return 0 or 1 message) that returns 1 message should create a `receive` span. Can you confirm ?

g7ed6e replied:
@lachmatt I kept the link but removed the parent/child relation

CodeBlanch commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,243 @@
+            return null;
+        }
+
+        activity.SetTag(SemanticConventions.AttributeMessagingSystem, MessagingSystem);
+        activity.SetTag(SemanticConventions.AttributeMessagingClientId, this.Name);
+        activity.SetTag(SemanticConventions.AttributeMessagingDestinationName, topic);
+        activity.SetTag(SemanticConventions.AttributeMessagingOperation, operation);
+
+        if (message.Key != null)
+        {
+            activity.SetTag(SemanticConventions.AttributeMessagingKafkaMessageKey, message.Key);
+        }
```
```suggestion
        if (activity.IsAllDataRequested)
        {
            activity.SetTag(SemanticConventions.AttributeMessagingSystem, MessagingSystem);
            activity.SetTag(SemanticConventions.AttributeMessagingClientId, this.Name);
            activity.SetTag(SemanticConventions.AttributeMessagingDestinationName, topic);
            activity.SetTag(SemanticConventions.AttributeMessagingOperation, operation);

            if (message.Key != null)
            {
                activity.SetTag(SemanticConventions.AttributeMessagingKafkaMessageKey, message.Key);
            }
        }
```

CodeBlanch commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,243 @@
+            return null;
+        }
+
+        activity.SetTag(SemanticConventions.AttributeMessagingKafkaDestinationPartition, topicPartition.Partition.Value);
```
```suggestion
        if (activity.IsAllDataRequested)
        {
            activity.SetTag(SemanticConventions.AttributeMessagingKafkaDestinationPartition, topicPartition.Partition.Value);
        }

```

CodeBlanch commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedConsumer.cs:
```diff
@@ -0,0 +1,326 @@
+            : Array.Empty<ActivityLink>();
+
+        Activity? activity = ConfluentKafkaCommon.ActivitySource.StartActivity(spanName, kind: ActivityKind.Consumer, links: activityLinks, startTime: start, parentContext: default);
+        if (activity != null)
```
```suggestion
        if (activity?.IsAllDataRequested == true)
```

CodeBlanch reviewed (commented):
Suggestions to check `Activity.IsAllDataRequested` before adding tags.

CodeBlanch commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,243 @@
+
+    private void InjectActivity(Activity? activity, Message<TKey, TValue> message)
+    {
+        (ActivityContext activityContext, Baggage baggage) = activity != null
+            ? (activity.Context, Baggage.Current.SetBaggage(activity.Baggage))
+            : (default, default);
```
```suggestion
        (ActivityContext activityContext, Baggage baggage) = activity != null
            ? (activity.Context, Baggage.Current)
            : (default, default);
```

This seems to be trying to stuff `Activity.Baggage` into OTel `Baggage`. That _probably_ is a bad idea because AFAIK no other instrumentation does that. We generally recommend users do not use `Activity.Baggage` at all.

/cc @cijothomas

CodeBlanch commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs:
```diff
@@ -0,0 +1,243 @@
+
+    private void InjectActivity(Activity? activity, Message<TKey, TValue> message)
+    {
+        (ActivityContext activityContext, Baggage baggage) = activity != null
+            ? (activity.Context, Baggage.Current.SetBaggage(activity.Baggage))
+            : (default, default);
+        this.propagator.Inject(new PropagationContext(activityContext, baggage), message, this.InjectTraceContext);
```
```suggestion
        this.propagator.Inject(new PropagationContext(activity?.Context ?? default, Baggage.Current), message, this.InjectTraceContext);
```

This seems to be trying to stuff `Activity.Baggage` into OTel `Baggage`. That _probably_ is a bad idea because AFAIK no other instrumentation does that. We generally recommend users do not use `Activity.Baggage` at all.

/cc @cijothomas

CodeBlanch replied:
(Edited original comment so that it is also possible to inject baggage without an activity.)

g7ed6e commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedConsumer.cs:
```diff
@@ -0,0 +1,272 @@
+    {
+        this.EnsurePreviousConsumeActivityIsStopped();
+
+        var consumeResult = this.consumerImplementation.Consume(cancellationToken);
```
done

g7ed6e commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedConsumer.cs:
```diff
@@ -0,0 +1,338 @@
+            : string.Concat(consumeResult.TopicPartitionOffset!.Topic, " ", OperationName);
+
+        using var activity = this.StartReceiveActivity(spanName, propagationContext, startTime, consumeResult.TopicPartitionOffset, consumeResult.Key);
+        activity?.Stop();
```
@lachmatt I just pushed the metering part. there are places where I try to pull the activity duration to push a metric thus the Stop call is required to get duration. That being said I will chase places where Stop calls can be removed.

CodeBlanch commented:
@g7ed6e I pushed some changes to try and get this over the finish line 😄 

* Missing `MinVerTagPrefix` in the `.proj`: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493/commits/f3c4df8b5e88dcee7899297011b71478f98c9148

* Static meters/counters: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493/commits/a4fc992e507bea855489b7034b8129a8ec0c6120

  The way it was before users could end up with multiple meters in their process with the same name & version. That is a no-no in the OTel spec. SDK should actually be fine with it but it seemed simpler to just make them static.

  Since I was in there I also switched the measurements to use `TagList` so they become allocation free. Before for every measurement it was allocating an enumerator and array. Expensive!

* Lit up the integration tests in CI: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493/commits/400fdc15ccc3521bac45b6fa64f873dd402c77b9

  The tests are currently failing. Would you please look into that?

@Kielek @vishweshbankwar Some concerns with the tests being added on this PR. There are no unit tests, only integration tests. Usually we have a mix of both? Not a blocker for me so long as the CI runs what is there but I think ideally there would be some unit tests too. Could be a follow-up effort.

g7ed6e commented:
> @g7ed6e I pushed some changes to try and get this over the finish line 😄
> 
> * Missing `MinVerTagPrefix` in the `.proj`: [f3c4df8](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/commit/f3c4df8b5e88dcee7899297011b71478f98c9148)
> * Static meters/counters: [a4fc992](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/commit/a4fc992e507bea855489b7034b8129a8ec0c6120)
>   The way it was before users could end up with multiple meters in their process with the same name & version. That is a no-no in the OTel spec. SDK should actually be fine with it but it seemed simpler to just make them static.
>   Since I was in there I also switched the measurements to use `TagList` so they become allocation free. Before for every measurement it was allocating an enumerator and array. Expensive!
> * Lit up the integration tests in CI: [400fdc1](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/commit/400fdc15ccc3521bac45b6fa64f873dd402c77b9)
>   The tests are currently failing. Would you please look into that?
> 
> @Kielek @vishweshbankwar Some concerns with the tests being added on this PR. There are no unit tests, only integration tests. Usually we have a mix of both? Not a blocker for me so long as the CI runs what is there but I think ideally there would be some unit tests too. Could be a follow-up effort.

@CodeBlanch I pulled your changes and updated the integration tests.

CodeBlanch commented:
@g7ed6e Seems like the integration tests are still having an issue: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/9683812113/job/26720211596?pr=1493

g7ed6e commented:
> @g7ed6e Seems like the integration tests are still having an issue: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/9683812113/job/26720211596?pr=1493

@CodeBlanch it looks like multiple Kafka containers are created. I don't face this issue locally I will check CI scripts.

CodeBlanch reviewed (approved):
I do still think we could do more to help connect traces when consuming (https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493#discussion_r1640199079) but LGTM that could be done as follow-up work.

g7ed6e commented:
> I do still think we could do more to help connect traces when consuming ([#1493 (comment)](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493#discussion_r1640199079)) but LGTM that could be done as follow-up work.

@CodeBlanch I missed that comment. I propose to implement it in a follow up PR.

CodeBlanch commented:
@Kielek @vishweshbankwar I think this is good to go but since it is touching a lot of the repo I would like another maintainer to approve before merging.

vishweshbankwar commented on build/Common.props:
```diff
@@ -43,6 +43,7 @@
     <OpenTelemetryCoreLatestVersion>[1.9.0,2.0)</OpenTelemetryCoreLatestVersion>
     <OpenTelemetryCoreLatestPrereleaseVersion>[1.9.0-rc.1]</OpenTelemetryCoreLatestPrereleaseVersion>
     <StackExchangeRedisPkgVer>[2.1.58,3.0)</StackExchangeRedisPkgVer>
+    <ConfluentKafkaPkgVer>[2.3.0,3.0)</ConfluentKafkaPkgVer>
```
not a blocker - this could be bump to 2.4.0 now?

g7ed6e replied:
@vishweshbankwar  Indeed 2.4.0 has been released since I started working on this. I will push the package update in one of the planned follow up PR.

vishweshbankwar commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/README.md:
````diff
@@ -0,0 +1,11 @@
+
+```shell
+dotnet add package OpenTelemetry.Instrumentation.ConfluentKafka --prerelease
+```
````
It would be good to link the example from here. or have some basic examples on how to configure the instrumentation. (Not a blocker - can be a follow up PR)

vishweshbankwar reviewed (approved):
Looks good to me overall. Left couple minor comments that can be addressed in the follow up.

I do have the same concern as @CodeBlanch noted in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493#issuecomment-2189854737 about the lack of unit tests. Given this PR is already big, it can be addressed in a follow up.

Kielek reviewed (approved):
Good to merge when you join to open-telemetry organization.

g7ed6e commented:
> Good to merge when you join to open-telemetry organization.

@Kielek  I joined the open-telemetry organization.

roseegalla commented:
@Kielek and @g7ed6e are there any plans to publish this package?

g7ed6e commented:
Sure. However this is not ready yet, there's still some work before publishing.

rgallagher-adc commented on src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetry.Instrumentation.ConfluentKafka.csproj:
```diff
@@ -0,0 +1,35 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>net8.0;net6.0;net462</TargetFrameworks>
```
@g7ed6e curious if there was a reason you didn't include .Net Standard 2.0? If not can it be added as a target framework?
